### PR TITLE
Use 31 bit limbs, do constant time and implement montgomery multiplication.

### PR DESCRIPTION
**Pull request by coltfred**
In order to better support WASM we moved the Fp types to use 31 bit limbs as it's the lowest common denominator for all platforms.

This turned out to be roughly an entire rewrite of this library. Most of it was ported from [BearSSL](https://www.bearssl.org/bigint.html), which is both constant time and very efficient. We updated some algorithms and tried to use better types where possible. There is still certainly work to do here, but this is a step in the right direction.

At this point in time, we believe that the entire library is constant time, though that has not been subject to audit, yet. The `Monty` type, which is generated in the module you specify when you invoke the macro is the montgomery form of the value. This form allows for a multiplication that is roughly twice as fast as normal. Division is not supported in montgomery form as the division algorithm we chose does not play well with it. I didn't research any further, though it might be possible in the future.

TODO
- [x] Update README
- [x] More comments.

### CODE REVIEW

File: src/digits/constant_bool.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    #[inline]
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self = *self | rhs;
+    }
+}
+
+//Neg<Output = Wrapping<T>>
+impl<T> ConstantBool<T>
+where
+    // u64: From<T>,
```
Comment (clintfred) [inline src/digits/constant_bool.rs]: remove?

File: src/digits/constant_bool.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        + Zero
+        + PartialEq,
+{
+    #[inline]
+    pub fn mux(self, x: T, y: T) -> T {
+        y ^ (Wrapping(self.0).neg().0 & (x ^ y))
+    }
+    #[inline]
+    pub fn is_zero(i: T) -> Self {
+        // let q = i as u64;
```
Comment (clintfred) [inline src/digits/constant_bool.rs]: remove

File: src/digits/ff31.rs [hunk 3]
```diff
@@ -0,0 +1,1373 @@
+///Convert the src into the limbs. This _does not_ mod off the value. This will take the first
+///len bytes and split them into 31 bit limbs.
+///Note that this will _not_ check anything about the length of limbs and could be unsafe... BE CAREFUL!
```
Comment (clintfred) [inline src/digits/ff31.rs]: What are the failure conditions? Should we add "unsafe" to the name of the function?

File: src/digits/ff31.rs [hunk 4]
```diff
@@ -0,0 +1,1373 @@
...
+            acc = b_u32 >> (8 - acc_len);
+        }
+    }
+    if acc_len != 0 {
+        limbs[v] = acc;
+        // v += 1
+    }
+}
+
+//COLT: Comment is out of date.
```
Comment (clintfred) [inline src/digits/ff31.rs]: Comment still out of date?

File: src/digits/ff31.rs [hunk 5]
```diff
@@ -0,0 +1,1373 @@
...
+            pub const MONTRSQUARED: [u32; NUMLIMBS] = $montgomery_r_squared;
+            pub const MONTM0INV: u32 = $montgomery_m0_inv;
+
+            #[derive(PartialEq, Eq, Ord, Clone, Copy)]
+            pub struct $classname {
+                pub(crate) limbs: [u32; NUMLIMBS],
+            }
+
+            #[derive(PartialEq, Eq, Ord, Clone, Copy)]
+            pub struct Monty {
```
Comment (clintfred) [inline src/digits/ff31.rs]: Worth noting that this in Montgomery form? And/or rename to `MontyForm`?

File: src/digits/ff31.rs [hunk 6]
```diff
@@ -0,0 +1,1373 @@
...
+                fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+                    write!(f, "{}(", stringify!($classname))?;
+                    let x: Vec<String> = self.limbs.iter().map(|x| format!("{:#x}", x)).collect();
+                    write!(f, "{}", x.join(", "))?;
+                    write!(f, ")")?;
+                    Ok(())
+                }
+            }
+
+            // impl fmt::Display for $classname {
```
Comment (clintfred) [inline src/digits/ff31.rs]: Not sure if this code works or not. If it does, is there any downside to uncommenting it?

File: src/digits/ff31.rs [hunk 7]
```diff
@@ -0,0 +1,1373 @@
...
+                pub fn to_norm(self) -> $classname {
+                    let mut one = [0u32; NUMLIMBS];
+                    one[0] = 1;
+                    $classname {
+                        limbs: (self * Monty { limbs: one }).limbs,
+                    }
+                }
+
+                #[inline]
+                pub fn normalize_assign_little(&mut self) {
```
Comment (clintfred) [inline src/digits/ff31.rs]: Is it worth describing "little" vs "big" limbs somewhere?

File: src/digits/ff31.rs [hunk 8]
```diff
@@ -0,0 +1,1373 @@
...
+                    let new_limbs = $classname::normalize_little_limbs(self.limbs);
+                    self.limbs = new_limbs;
+                }
+
+                ///Take the extra limb and incorporate that into the existing value by modding by the prime.
+                /// This normalize should only be used when the input is at most
+                /// 2*p-1. Anything that might be bigger should use the normalize_big
+                /// options, which use barrett.
```
Comment (clintfred) [inline src/digits/ff31.rs]: Would be nice to note somewhere the cost/benefit that was done of when to use Monty vs Barrett...

File: src/digits/ff31.rs [hunk 9]
```diff
@@ -0,0 +1,1373 @@
...
+                pub fn normalize_assign_big(&mut self, extra_limb: u32) {
+                    let mut ret = [0u32; NUMLIMBS * 2];
+                    for (dst, src) in ret.iter_mut().zip(self.limbs.iter()) {
+                        *dst = *src;
+                    }
+                    ret[NUMLIMBS] = extra_limb;
+                    self.limbs = $classname::reduce_barrett(&ret);
+                }
+
+                ///This function assumes that the buf pointer has at least 4 spaces starting at the beginning of the
```
Comment (clintfred) [inline src/digits/ff31.rs]: should we assert this, or similar?

File: src/digits/ff31.rs [hunk 10]
```diff
@@ -0,0 +1,1373 @@
...
+                    output
+                }
+
+                ///Create a new instance given the raw limbs form. Note that this is least significant bit first.
+                #[allow(dead_code)]
+                pub fn new(digits: [u32; NUMLIMBS]) -> $classname {
+                    $classname { limbs: digits }
+                }
+
+                ///Write out the value in decimal form.
```
Comment (clintfred) [inline src/digits/ff31.rs]: Would love to see this implemented. :)

File: src/lib.rs [hunk 11]
```diff
@@ -7,600 +9,648 @@ extern crate proptest;
...
 }
 
 #[cfg(test)]
 mod lib {
     use super::*;
-    use num_traits::{Inv, One, Pow, Zero};
-    use std::ops::{Div, Mul};
+    use num_traits::{One, Zero};
+
+    // #[test]
```
Comment (clintfred) [inline src/lib.rs]: Do these tests need to come back? Get deleted?

File: src/lib.rs [hunk 12]
```diff
@@ -7,600 +9,648 @@ extern crate proptest;
...
-            17072191670499601912,
-            15567716623030039384,
-            10534201835020433733,
-            2741051863,
-        ]);
-        assert_eq!(a * b, c);
-        assert_eq!(c / b, a);
-        assert_eq!(c / a, b);
-    }
+    // #[test]
```
Comment (clintfred) [inline src/lib.rs]: Commented tests

File: src/lib.rs [hunk 13]
```diff
@@ -612,201 +662,174 @@ mod lib {
...
-            max.normalize(1),
-            fp_256::Fp256::new([
-                13180341465104399562,
-                3813816114071133989,
-                49603190868503450,
-                5827933165629827091
-            ])
-        );
-    }
+    // #[test]
```
Comment (clintfred) [inline src/lib.rs]: commented tests

File: src/digits/ff31.rs [hunk 14]
```diff
@@ -212,30 +186,31 @@ macro_rules! fp31 {
...
                 #[inline]
-                fn mul(mut self, rhs: u32) -> $classname {
-                    self *= rhs.into();
-                    self
+                fn mul(self, rhs: u32) -> $classname {
+                    util::sum_n(self, rhs as u64)
                 }
             }
-
+            ///Note that this reveals the u32, but nothing else. It's expected that the u32 is not secret.
```
Comment (clintfred) [inline src/digits/ff31.rs]: This should be `u64`

File: src/digits/ff31.rs [hunk 15]
```diff
@@ -428,6 +399,26 @@ macro_rules! fp31 {
                 }
             }
 
+            ///Note that this reveals the u32, but nothing else. It's expected that the u32 is not secret.
```
Comment (clintfred) [inline src/digits/ff31.rs]: u64

File: src/lib.rs (hunk 11 above)
Comment (clintfred) [reply src/lib.rs]: Your comment says your removed the test, but it's still here.

File: src/lib.rs (hunk 12 above)
Comment (clintfred) [reply src/lib.rs]: Removed?

File: src/digits/ff31.rs (hunk 14 above)
Comment (coltfred) [reply src/digits/ff31.rs]: fixed.

File: src/digits/ff31.rs (hunk 15 above)
Comment (coltfred) [reply src/digits/ff31.rs]: fixed.

**Pull request merged**